core/node/core.go

```diff
@@ -168,3 +169,54 @@ func Files(mctx helpers.MetricsCtx, lc fx.Lifecycle, repo repo.Repo, dag format.
 
 	return root, err
 }
+
+// FIXME(BLOCKING): This should probably be merged into the Files provider above
+//  as no one should have access to the unlocked version of the mfs.Root.
+func LockedFiles(mctx helpers.MetricsCtx, lc fx.Lifecycle, root *mfs.Root) (*LockedFilesRoot, error) {
+	return &LockedFilesRoot{
+		root: root,
+		lock: &sync.Mutex{},
+	}, nil
+}
+
+// We distinguish between locking for reading and writing independently of
+// how is the lock actually implemented internally (we could be write-locking
+// every time, using a RWMutex, or not even locking for reading).
+// The objective is:
+// > For reads we might want to be more careful than a global lock though.
+// > For example, not locking ipfs files ls or ipfs files stat while GC is ongoing,
+// > but preventing corruption between MFS write and read operations (out of date data
+// > is fine with concurrent operations, but corruption would be bad).
+// FIXME(BLOCKING): Decide how are we actually going to implement the locking
+//  distinction between reading and writing.
+// FIXME(BLOCKING): Consider providing our own API functions like GetNode (and similar) that
+//  already take charge of lock-handling here instead of leaving that burden to the consumer.
+//  That way we could also make sure the lock is being enforced correctly.
```

Comment:
I would leave things exactly as you have them right now. There are bigger issues with MFS locking that cannot be resolved by any sort of locking on the root alone. This PR is trying to just fix races between GC and competing command-line MFS operations.

Reply:
I like the interface of LockedFilesRoot, because having only RLock and Lock methods forces whoever is touching this code to think/ask what is going on.

Reply:
Yes, sorry, this is actually a non-blocking fixme, just a note of a possible direction down the road, will clarify that.